### PR TITLE
refactor KeyRange and support from-key

### PR DESCRIPTION
**Pull request by themanforfree**

### CODE REVIEW

File: xline/src/server/command.rs [hunk 1]
```diff
@@ -20,11 +17,29 @@ use tokio::sync::oneshot;
...
 use crate::storage::KvStore;
 
-/// Range end to get all keys
-const ALL_KEYS: &[u8] = &[0_u8];
+/// Range start and end to get all keys
+const UNBOUNDED: &[u8] = &[0_u8];
 /// Range end to get one key
 const ONE_KEY: &[u8] = &[];
 
+pub(crate) enum RangeType {
```
Comment (pzheng1025) [inline xline/src/server/command.rs:25]: Add docs.

File: xline/src/server/command.rs [hunk 2]
```diff
@@ -37,39 +52,66 @@ pub(crate) struct KeyRange {
...
+            (Bound::Unbounded, Bound::Unbounded) => return true,
+            _ => unreachable!("KeyRange::start_bound() cannot be Excluded"),
+        } {
+            // s1 < s2
+            // s2 < e1 ?
+            match (other.start_bound(), self.end_bound()) {
+                (Bound::Included(s2), Bound::Included(e1)) => s2 <= e1,
+                (Bound::Included(s2), Bound::Excluded(e1)) => s2 < e1,
+                (Bound::Included(_), Bound::Unbounded) => true,
+                // if other.start_bound() is Unbounded, programe cannnot enter this branch
```
Comment (pzheng1025) [inline xline/src/server/command.rs:74]: typo `cannot`

File: xline/src/server/command.rs [hunk 3]
```diff
@@ -37,39 +52,66 @@ pub(crate) struct KeyRange {
...
     pub(crate) fn contains_key(&self, key: &[u8]) -> bool {
-        match self.end.as_slice() {
-            ONE_KEY => self.start == key,
-            ALL_KEYS => true,
-            _ => (key >= self.start.as_slice()) && (key < self.end.as_slice()),
-        }
+        self.contains(key)
+    }
+
+    pub(crate) fn range_type(&self) -> RangeType {
```
Comment (pzheng1025) [inline xline/src/server/command.rs:97]: Add docs. Do you know why clippy doesn't catch this?

File: xline/src/server/command.rs [hunk 4]
```diff
@@ -20,11 +17,29 @@ use tokio::sync::oneshot;
...
 const ONE_KEY: &[u8] = &[];
 
+pub(crate) enum RangeType {
+    OneKey,
+    AllKeys,
+    Range,
+}
+
+impl RangeType {
+    pub(crate) fn get_range_type(key: &[u8], range_end: &[u8]) -> Self {
```
Comment (pzheng1025) [inline xline/src/server/command.rs:32]: Add docs

File: xline/src/server/command.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/server/command.rs:103]: fixed in #50

**Review by pzheng1025 — APPROVED** [review]

**Pull request merged**